test/common/ext_authz/ext_authz_impl_test.cc

```diff
@@ -0,0 +1,193 @@
+    EXPECT_CALL(request_callbacks_, onComplete(CheckStatus::Denied));
+    client_.onSuccess(std::move(response), span_);
+  }
+
+  {
+    envoy::service::auth::v2::CheckRequest request;
+    Http::HeaderMapImpl headers;
+    EXPECT_CALL(*async_client_, send(_, ProtoEq(request), _, _, _))
+        .WillOnce(Return(&async_request_));
+
+    client_.check(request_callbacks_, request, Tracing::NullSpan::instance());
+
+    client_.onCreateInitialMetadata(headers);
+
+    response.reset(new envoy::service::auth::v2::CheckResponse());
+    ::google::rpc::Status* status = new ::google::rpc::Status();
+    status->set_code(Grpc::Status::GrpcStatus::Ok);
+    response->set_allocated_status(status);
+    EXPECT_CALL(span_, setTag("ext_authz_status", "ext_authz_ok"));
+    EXPECT_CALL(request_callbacks_, onComplete(CheckStatus::OK));
+    client_.onSuccess(std::move(response), span_);
+  }
+
+  {
+    envoy::service::auth::v2::CheckRequest request;
+    EXPECT_CALL(*async_client_, send(_, ProtoEq(request), _, _, _))
+        .WillOnce(Return(&async_request_));
+
+    client_.check(request_callbacks_, request, Tracing::NullSpan::instance());
+
+    response.reset(new envoy::service::auth::v2::CheckResponse());
+    EXPECT_CALL(request_callbacks_, onComplete(CheckStatus::Error));
+    client_.onFailure(Grpc::Status::Unknown, "", span_);
+  }
+}
+
+TEST_F(ExtAuthzGrpcClientTest, Cancel) {
+  std::unique_ptr<envoy::service::auth::v2::CheckResponse> response;
+  envoy::service::auth::v2::CheckRequest request;
+
+  EXPECT_CALL(*async_client_, send(_, _, _, _, _)).WillOnce(Return(&async_request_));
+
+  client_.check(request_callbacks_, request, Tracing::NullSpan::instance());
+
+  EXPECT_CALL(async_request_, cancel());
+  client_.cancel();
+}
+
+class CreateCheckRequestTest : public testing::Test {
+public:
+  CreateCheckRequestTest() {
+    addr_ = std::make_shared<Network::Address::Ipv4Instance>("1.2.3.4", 1111);
+    protocol_ = Envoy::Http::Protocol::Http10;
+  };
+
+  Network::Address::InstanceConstSharedPtr addr_;
+  Optional<Http::Protocol> protocol_;
+  CreateCheckRequest check_request_generator_;
+  NiceMock<Envoy::Http::MockStreamDecoderFilterCallbacks> callbacks_;
+  NiceMock<Envoy::Network::MockReadFilterCallbacks> net_callbacks_;
+  NiceMock<Envoy::Network::MockConnection> connection_;
+  NiceMock<Envoy::Ssl::MockConnection> ssl_;
+  NiceMock<Envoy::RequestInfo::MockRequestInfo> req_info_;
+};
+
+TEST_F(CreateCheckRequestTest, BasicTcp) {
+
+  envoy::service::auth::v2::CheckRequest request;
+
+  EXPECT_CALL(net_callbacks_, connection()).Times(2).WillRepeatedly(ReturnRef(connection_));
+  EXPECT_CALL(connection_, remoteAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(connection_, localAddress()).WillOnce(ReturnRef(addr_));
+  EXPECT_CALL(Const(connection_), ssl()).Times(2).WillRepeatedly(Return(&ssl_));
+
+  CreateCheckRequest::createTcpCheck(&net_callbacks_, request);
+}
+
+TEST_F(CreateCheckRequestTest, BasicHttp) {
```

Comment:
https://23972-65214191-gh.circle-artifacts.com/0/build/envoy/generated/coverage/coverage.source_common_ext_authz_ext_authz_impl.cc.html tells me you have full coverage here, but I'm a bit surprised looking at just these tests, since they don't seem to be driving all the paths (at least directly). Could you say a few words on why we get full coverage here in this thread?

Reply:
@htuch 
The underlying structures like `Ssl::Connection` context are mocks hence they are there but empty. The calls from it here thus invoks all the conditional code.
The other connditional is around `local` boolean which invokes with both `true` and `false` within the `createTcpCheck` and `createHttpCheck` it self.
The remaining conditional checks like protocol and service is invoked because those structures are getting initialized in the tests.

I've addressed the rest of the feedback as well. Thanks for the review...it has made the code much better.